### DISCUSSION TITLE
incremental parser support

### DISCUSSION THREAD

**Discussion by IWANABETHATGUY** [vuejs/language-tools#58] [General]
tree-sitter is a error tolerant and incremental parser, which is blazing fast for editor scenario, i took a glance at our codebase, which has using a lot of Regular Expression and hardcoding parsing, which is  inefficient. I am pretty familiar with tree-sitter , and I am willing to integrate it with our project. However i am not familiar with our project ,  maybe you could write a more detailed infrastructure to me ?

**Comment by IWANABETHATGUY** [comment]
here is the `tree-sitter` homepage: https://tree-sitter.github.io/tree-sitter/
`tree-sitter-vue` : https://ikatyang.github.io/tree-sitter-vue/ you can try here,
![8UWKyACuVU](https://user-images.githubusercontent.com/17974631/106875027-2eab1100-6711-11eb-93ba-3ae042638dae.gif)
1200+ line `vue` file , first parsing only took `6ms`, after that , every keystroke only cost `0.3` ms

**Comment by johnsoncodehk** [comment]
Yes, of course we can consider using it :) Solving performance problems is the purpose of establishing this project.

Quick preview for what volar parser using:

|                         | parser                                                        | for                                            |
| ----------------------- | ------------------------------------------------------------- | ---------------------------------------------- |
| SFC                     | @vue/compiler-sfc -> parse                                    | Virtual documents mapping                      |
| `<template>`            | @vue/compiler-dom -> compile                                  | Virtual template script                        |
| `<template lang="pug">` | @vue/compiler-dom -> compile + ~~**Regex**~~ pug-parser                      | Virtual template script + html <-> pug mapping |
| `<script>`              | typescript -> createSourceFile -> getStart / End              | Virtual `<script setup>` script                |
| `<script setup>`        | typescript -> createSourceFile -> getStart / End              | Virtual `<script setup>` script                |
| `<style>`               | vscode-css-languageservice -> findDocumentSymbols + **Regex** | Extra css service                              |

There are currently no obvious performance problems with regex, because the usage of regex is very small, and in the case of on-demand calculation, regex will not be calculated if `<template lang="pug">`, `<style>` block is not changed.

But we have performance problems from typescript's `getStart()`, `getStart()` is very expensive, and the diagnosis will be significantly lagging when ref sugar is used extensively in `<script setup>`. Replacing typescript with tree-sitter seems to solve this problem.

The project is still changing rapidly, discussing efficiency in github may be very low, you can find me on discord if you need, my id is `johnsoncodehk#0495`.

**Reply by cool-delete** [reply]
@johnsoncodehk

Is there a shortcut to quickly indent PUG?
example
```
p
  span()
    | bbac。
  br
  span()
    | cdeee
```
How can I get off the P tag quickly?

**Reply by johnsoncodehk** [reply]
@cool-delete do you mean remove `() \n |` with formatting? Formatter don't handle it, but you could use volar's pug converter(click `html ☐` and then click `pug ☐`), `()` will be remove remove because it don't have attributes.

**Reply by cool-delete** [reply]
no  I need to
```
p
  span()
    | bbac。
  br
  span()
    | cdeee

```
become

```

span
  | bbac。
br
span()
  | cdeee
```

**Reply by cool-delete** [reply]
If I remove the p tag directly, there will be an error indicating invalid indentation.

**Reply by johnsoncodehk** [reply]
> If I remove the p tag directly, there will be an error indicating invalid indentation.

This is the correct report. pug is sensitive to indentation, the tool will not and should not change the indentation of the root node, because this will produce different compilation results. You can use shift+tab to downgrade the indentation of a piece of code in vscode.

**Reply by cool-delete** [reply]
` shift+tab` work for me!

**Comment by johnsoncodehk** [comment]
Please track #1718.

**Reply by johnsoncodehk** [reply]
Implemented in https://github.com/johnsoncodehk/volar/commit/7337d1c1fef9f51c70387e9352e7e4914800ea2d and https://github.com/johnsoncodehk/volar/commit/fbc6971ab09c6b7ad1d81b16f31fed855037fc6c.
